release-24.1: roachtest/mixedversion: add support for shared-process deployments

renatolabs opened the pull request:
Backport:
  * 2/2 commits from "roachtest/mixedversion: add support for shared-process deployments" (#127393)
  * 2/2 commits from "roachtest: improvements in connection to mixedversion services" (#127900)
  * 1/1 commits from "roachtest: increase internal query timeout in mixedversion" (#128186)
  * 1/1 commits from "roachtest: change-replicas/mixed-version supports shared-process" (#128119)

Please see individual PRs for details.

/cc @cockroachdb/release

Release justification: test only changes.

Commit:
roachtest/mixedversion: add support for shared-process deployments

Up to this point, tests that use the `mixedversion` framework exercise
upgrade paths related to the system (storage) cluster exclusively.
While, technically, they could create their own tenants and upgrade
them manually, that approach is not ideal as it requires discipline
from every test to implement.

In this commit, we introduce support for the framework itself to
create a multi-tenant deployment out of the box. The test will set the
`default_target_cluster` to the tenant created, enabling individual
tests to run on a multi-tenant deployment without being aware of it.

Note that we also disable the new deployment mode in a number of
existing mixedversion tests in this commit. These tests need (often
minor) tweaks to work in a multi-tenant environment. Fixes for these
tests will come in follow-up commits.

Epic: none

Release note: None

Commit:
roachtest: change-replicas/mixed-version supports shared-process

This commit updates the `change-replicas/mixed-version` roachtest in
order to enable the shared-process deployment mode. This involves
enabling certain features in tenants (such as the ability to split and
scatter) in versions where they are not enabled by default. One of the
builtins used also require a connection to the system tenant.

Informs: #127378

Release note: None

Commit:
roachtest: print mixedversion seed along with test plan

Previously, the seed would be printed as a regular test log line,
which was slightly harder to find in the test logs. With this change,
the seed is displayed in the test plan itself, before the list of
upgrades performed.

Epic: none

Release note: None

Commit:
roachtest/mixedversion: limit max time spent in internal queries

In this commit, we change the `BinaryVersion` and `ClusterVersion`
functions accept a `context` parameter. In addition, we change the
mixedversion calls to these functions to pass a context with a fixed
timeout. This effectively limits the maximum amount of time we will
spend trying to gather this debug information.

In some failure scenarios, nodes may become unresponsive, which may
make the test seem "stuck." With this change, tests should fail if,
for whatever reason, we can't determine a node's binary or cluster
version within 10s.

Epic: none

Release note: None

Commit:
roachtest: make sure tenant is ready after restart in mixedversion

This updates the start and restart steps in the mixedversion framework
to wait for the shared-process tenant (if any) to be able to handle
incoming requests before moving on, as the test plan generated by the
framework assumes that the tenant is able to handle connections after
the `start` step.

In rare occasions, it is possible for the tenant to have started but, by
the time the framework attempts to query internal debug information
from the tenant, the tenant is not ready to accept requests yet.

We retry for a short period of time until we are sure the tenant
server is up.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/128120)
<!-- Reviewable:end -->

Commit:
roachtest: increase internal query timeout in mixedversion

Previously, this was set to 10s. However, in a few runs, it took more
than 10s for the server to initialize after a restart. In this commit,
we increase the timeout to 30s and also query the internal debug
information concurrently which should make a substantial difference in
larger clusters.

Epic: none

Release note: None

Commit:
roachtest: release-24.1 fixes for mixedversion shared-process

This applies fixes that are specific to tests in this release branch.

Epic: none

Release note: None

renatolabs commented:
TFTR!
